### PR TITLE
Create Gauge APR Provider

### PR DESCRIPTION
**Pull request by hammeiam**
- move `getPriceDataForPool` from `usePoolData` to `utils` for easier sharing
- Create `AprsProvider` to calculate gauge APRs. These are represented two ways: `apr` for rewards that we know the price of, and `amountPerDay` for rewards that we don't know the price of (SDL)
- Add `AprsProvider` to App

Math:
Per John's Notion notes, `Reward.rate` is amount per second and represents the max award rate. Thus the min rate is `Reward.rate * 0.4`. The following formula is used to calculate Min APR
`0.4 * (Reward.rate * gauge.relative_weight) * year_secs * rewardPriceUSD / gauge.working_supply * lpTokenPriceUSD`

Screen Shot 2022-06-17 at 3 09 33 PM

Screen Shot 2022-06-17 at 5 09 45 PM

### CODE REVIEW

File: src/utils/index.ts [hunk 1]
```diff
@@ -491,3 +493,52 @@ export function extractAddEthereumChainArgs(
...
+  tokenBalances1e18: BigNumber[]
+  totalLocked: BigNumber
+} {
+  const poolAssetPrice = parseUnits(
+    String(
+      tokenPricesUSD?.[getTokenSymbolForPoolType(basicPool.typeOfAsset)] || 0,
+    ),
+    18,
+  )
+  const expandedTokens = basicPool.tokens.map((token) => (tokens || {})[token])
```
Comment (ug02fast) [inline src/utils/index.ts:522]: {}[token] is going to be `undefined`. Is that ok?

**Review by ug02fast — APPROVED** [review]

**Comment by penandlim** [comment]
Correction on the APR calculation math. SDL rate is dictated by `Minter.rate() * gauge weight / 1e18` while external rewards are managed per gauge `gauge.reward_data[gauge.reward_tokens(i)].rate`



- Max and min SDL reward per Gauge (per second rate)

```jsx
Min: 0.4 * SDLPriceUSD * Minter.rate() * (GaugeController.gauge_relative_weight / 1e18) / (gauge.working_supply * lpTokenPriceUSD))
Max: SDLPriceUSD * Minter.rate() * (GaugeController.gauge_relative_weight / 1e18) / (gauge.working_supply * lpTokenPriceUSD))
```

- Each partner reward per Gauge (per second rate)

```jsx
Min: 0.4 * Reward[i].rate * rewardTokenPriceUSD / (gauge.working_supply * lpTokenPriceUSD)
Max: Reward[i].rate * rewardTokenPriceUSD / (gauge.working_supply * lpTokenPriceUSD)
```

File: src/providers/AprsProvider.tsx [hunk 2]
```diff
@@ -15,13 +15,18 @@ type MinAndMax = {
...
 type AmountReward = {
-  amount: MinAndMax
+  rewardToken: BasicToken
+  amountPerDay: MinAndMax
+  apr?: undefined
 }
 type AprReward = {
+  rewardToken: BasicToken
   apr: MinAndMax
+  amountPerDay?: undefined
```
Comment (ug02fast) [inline src/providers/AprsProvider.tsx:25]: Should this be `number` or `string`?

File: src/providers/AprsProvider.tsx [hunk 3]
```diff
@@ -15,13 +15,18 @@ type MinAndMax = {
   max: BigNumber
 }
 type AmountReward = {
-  amount: MinAndMax
+  rewardToken: BasicToken
+  amountPerDay: MinAndMax
+  apr?: undefined
```
Comment (ug02fast) [inline src/providers/AprsProvider.tsx:20]: Should this be `number` or `string`?

File: src/pages/VeSDL/LockedInfo.tsx [hunk 4]
```diff
@@ -60,10 +60,10 @@ export default function LockedInfo(): JSX.Element {
           {aggSDLInfo ? formatBNToShortString(aggSDLInfo.totalVeSDL, 18) : "-"}
         </Typography>
       </Box>
-      <Box flex={1}>
+      {/* <Box flex={1}>
```
Comment (ug02fast) [inline src/pages/VeSDL/LockedInfo.tsx:63]: Plans with this?

File: src/hooks/usePoolData.ts [hunk 5]
```diff
@@ -136,6 +147,7 @@ export default function usePoolData(poolName?: string): PoolDataHookReturnType {
       }
       try {
         const poolMinichefData = minichefData?.pools?.[basicPool.poolAddress]
+        // const poolGaugeData = gauges?.[basicPool.poolAddress]
```
Comment (ug02fast) [inline src/hooks/usePoolData.ts:150]: looks like you moved this lower in context, delete?

File: src/components/TokenClaimDialog.tsx [hunk 6]
```diff
@@ -301,29 +328,45 @@ export default function TokenClaimDialog({
...
       <Typography sx={{ flex: 1 }}>
-        {status === STATUSES.SUCCESS ? 0 : formattedAmount}
+        {title && <br />}
+        {items.map(([, amount]) => (
+          <>
+            {status === STATUSES.SUCCESS
+              ? 0
+              : commify(formatBNToString(amount, 18, 2))}
+            <br />
+          </>
```
Comment (ug02fast) [inline src/components/TokenClaimDialog.tsx:368]: Something like `<Typography mb={2}>// ternary</Typography>` could also work but this is accomplishing similar.

File: src/components/TokenClaimDialog.tsx [hunk 7]
```diff
@@ -301,29 +328,45 @@ export default function TokenClaimDialog({
...
+        )}
+        {items.map(([name]) => (
+          <>
+            {name}
+            <br />
+          </>
+        ))}
       </Typography>
       <Typography sx={{ flex: 1 }}>
```
Comment (ug02fast) [inline src/components/TokenClaimDialog.tsx:360]: This could also be `<Box sx={{...`

File: src/components/PoolOverview.tsx [hunk 8]
```diff
@@ -270,3 +250,81 @@ const StyledGrid = styled(Grid)<{ disabled?: boolean }>(
...
+          [key]: formatBNToPercentString(apr, 18),
+        }
+      : acc
+  }, {} as Partial<Record<string, string>>)
+  return (
+    <>
+      {Object.keys(poolData.aprs).map((key) => {
+        const symbol = poolData.aprs[key]?.symbol as string
+        return poolData.aprs[key]?.apr.gt(Zero) ? (
+          <div key={symbol}>
```
Comment (ug02fast) [inline src/components/PoolOverview.tsx:269]: Could also use Box

**Review by ug02fast — APPROVED** [review]
lgtm. added suggestions but nothing is breaking

File: src/components/TokenClaimDialog.tsx [hunk 9]
```diff
@@ -229,12 +238,13 @@ export default function TokenClaimDialog({
               </Trans>
             </Typography>
           )}
-          {!IS_VESDL_LIVE && chainId === ChainId.MAINNET
+          {!gaugesAreActive && chainId === ChainId.MAINNET
```
Comment (nguyenaiden) [inline src/components/TokenClaimDialog.tsx:241]: I think you can get rid of `chainId === ChainId.MAINNET` since `gaugeAreActive` already took care of that

**Review by nguyenaiden — APPROVED** [review]
LGTM after those comments.

File: src/pages/VeSDL/LockedInfo.tsx (hunk 4 above)
Comment (hammeiam) [reply src/pages/VeSDL/LockedInfo.tsx:63]: https://github.com/saddle-finance/saddle-frontend/issues/1092

File: src/providers/AprsProvider.tsx (hunk 3 above)
Comment (hammeiam) [reply src/providers/AprsProvider.tsx:20]: nope these are complementary types - one defines what the other leaves undefined. It allows us to both be unique and do `if (reward.apr) {...} else { reward.amount.min... }` while also having an overlapping type `RewardAmount | RewardApr`

File: src/components/PoolOverview.tsx (hunk 8 above)
Comment (hammeiam) [reply src/components/PoolOverview.tsx:269]: Good call! Sorry, I'm old fashioned

**Comment by hammeiam** [comment]
> Correction on the APR calculation math. SDL rate is dictated by `Minter.rate() * gauge weight / 1e18` while external rewards are managed per gauge `gauge.reward_data[gauge.reward_tokens(i)].rate`
> 
>     * Max and min SDL reward per Gauge (per second rate)
> 
> 
> ```js
> Min: 0.4 * SDLPriceUSD * Minter.rate() * (GaugeController.gauge_relative_weight / 1e18) / (gauge.working_supply * lpTokenPriceUSD))
> Max: SDLPriceUSD * Minter.rate() * (GaugeController.gauge_relative_weight / 1e18) / (gauge.working_supply * lpTokenPriceUSD))
> ```
> 
>     * Each partner reward per Gauge (per second rate)
> 
> 
> ```js
> Min: 0.4 * Reward[i].rate * rewardTokenPriceUSD / (gauge.working_supply * lpTokenPriceUSD)
> Max: Reward[i].rate * rewardTokenPriceUSD / (gauge.working_supply * lpTokenPriceUSD)
> ```

Thank you! I believe I've already captured the SDL rate here https://github.com/saddle-finance/saddle-frontend/blob/master/src/utils/gauges.ts#L150 - after the initial `minter.rate * relative_weight / 1e18` then I can do the same apr calculation as I'm doing with any other reward.

File: src/components/TokenClaimDialog.tsx (hunk 6 above)
Comment (hammeiam) [reply src/components/TokenClaimDialog.tsx:368]: I have this a shot, but I think that line breaks are probably a simpler way of creating a fake table, which is what I'm trying to do here. This UI could definitely be refined

**Comment by hammeiam** [comment]
@sandrasong please see screenshots above. Do you have any thoughts on how we could slightly improve the look of the new min/max ranges for aprs (eg `SDL APR: 5.50%-10.31%`) and amounts (eg `SDL/24h: 13,000-18,000`). We show an APR when we know the price of the reward token, and we only display the amount when we don't have a price.

**Pull request merged**